src/transforms/kubernetes/mod.rs

```diff
@@ -9,39 +9,69 @@ use crate::{
 };
 use bytes::Bytes;
 use evmap::{ReadHandle, WriteHandle};
-use futures::stream::Stream;
-use futures03::compat::Future01CompatExt;
+use futures::{
+    compat::{Future01CompatExt, Stream01CompatExt},
+    stream::StreamExt,
+};
 use k8s_openapi::api::core::v1::Pod;
+use rand::Rng;
 use serde::{Deserialize, Serialize};
 use snafu::{ResultExt, Snafu};
-use std::collections::BTreeMap;
-use std::collections::HashMap;
-use std::collections::VecDeque;
+use std::collections::{BTreeMap, VecDeque};
 use std::time::{Duration, Instant};
 use string_cache::DefaultAtom as Atom;
 use tokio::timer::Delay;
 
 // *********************** Defined by Vector **************************** //
+
 /// Node name `spec.nodeName` of Vector pod passed down with Downward API.
 const NODE_NAME_ENV: &str = "VECTOR_NODE_NAME";
 
-/// If watcher errors, for how long will we wait before trying again.
-const RETRY_TIMEOUT: Duration = Duration::from_secs(2);
-
-/// For how long will we hold on to pod's metadata after the pod has been deleted.
-///
-/// 10min sounds a lot, but a metadata entry averages around 300B and in an extreme
-/// scenario when a pod is deleted every second, we would have ~200kB of probably
-/// not usefull data. Which is acceptable. The benefit of such a long delay
-/// is that we will certainly* have the metadata while we are processing the
-/// remaining logs from the deleted pod.
-const DELETE_DELAY: Duration = Duration::from_secs(10 * 60);
-
 #[derive(Deserialize, Serialize, Debug)]
 #[serde(deny_unknown_fields)]
 pub struct KubePodMetadata {
     #[serde(default = "default_fields")]
-    fields: Vec<String>,
+    fields: Vec<Field>,
+    /// For how long will we hold on to pod's metadata after the pod has been deleted.
+    /// seconds
+    #[serde(default = "default_cache_ttl")]
+    cache_ttl: u64,
```

Comment:
nit we've been using `Option` to mark a field as optional then setting the default when we need to unwrap it.